### ISSUE TITLE
einsum wrong result / crash (migrated from Trac #1885)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3433]
Original ticket http://projects.scipy.org/numpy/ticket/1885
Reported 2011-06-28 by trac user royf, assigned to unknown.

numpy.einsum sometimes returns wrong results.
Consider the following example:

> > > p = numpy.array([.5, .5])
> > > q = numpy.array([[0.26894142, 1-0.26894142], [1-0.26894142, 0.26894142]])
> > > r = numpy.identity(2)
> > > ...
> > > p, q, r
> > > (array([ 0.5,  0.5]), array([[ 0.26894142,  0.73105858],
> > >        [ 0.73105858,  0.26894142]]), array([[ 1.,  0.],
> > >        [ 0.,  1.]]))
> > > numpy.einsum('z,mz,zm->', p, q, r)
> > > 3.7305612660634556e+177
> > > numpy.einsum('z,mz,zm->', p, q, r)
> > > 0.26894141999999999
> > > numpy.einsum('z,mz,zm->', p, q, r)
> > > 3.7305612660634556e+177
> > > numpy.einsum('z,mz,zm->', p, q, r)
> > > 0.26894141999999999
> > > numpy.einsum('z,mz,zm->', p, q, r)
> > > 3.7305612660634556e+177
> > > numpy.einsum('z,mz,zm->', p, q, r)
> > > 0.26894141999999999
> > > numpy.einsum('z,mz,zm->', p, q, r)
> > > 0.26894141999999999
> > > numpy.einsum('z,mz,zm->', p, q, r)
> > > 0.13447070999999999

In other runs this same example crashes the process.

The correct answer is 0.26894142.
Interestingly, one of the answers returned is half the correct one.
This probably has to do with the specific value of p in the example.
My guess is it's something like an off-by-one bug.

I'd try to fix it myself,
but the code is a scary mess :(

Good luck!

**Comment by thouis** [comment]
Comment in Trac by trac user royf, 2011-06-28

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2011-06-29

Fixed in commit:192ac749 (master) and commit:4362bdb (1.6.1 branch).

I backported the fix right away because it's a serious bug.

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2011-06-29

**Comment by thouis** [comment]
Comment in Trac by trac user royf, 2011-11-25

Replying to [comment:3 mwiebe]:

Problem not solved in 1.6.1,
or a very similar problem remains:

> > > import numpy; print(numpy.version.full_version); a = numpy.arange(1, 3); b = numpy.arange(1, 5).reshape(2, 2); c = numpy.arange(1, 9).reshape(4, 2); numpy.einsum('x,yx,zx->xzy', a, b, c)
> > > 1.6.1
> > > array([[[ 1,  3],
> > >         [ 3,  9],
> > >         [ 1,  3],
> > >         [ 3,  9]],

```
   [[ 8, 16],
    [16, 32],
    [ 8, 16],
    [16, 32]]])
```

The correct result is

> > > numpy.array([[[a[x]_b[y, x]_c[z, x] for y in range(2)] for z in range(4)] for x in range(2)])
> > > array([[[ 1,  3],
> > >         [ 3,  9],
> > >         [ 5, 15],
> > >         [ 7, 21]],

```
   [[ 8, 16],
    [16, 32],
    [24, 48],
    [32, 64]]])
```

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2012-01-21

I've created a ticket #2027 for the reopened version of the issue.
